packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -11,159 +11,240 @@
+  // unfortunately, finding the safe max/min is non-trivial, so we'll need to rely on people setting max/min correctly for their step size
+  // we can include it for percent though
+  if (intlOptions.style === 'percent' && isNaN(step)) {
+    maxValue = isNaN(props.maxValue) ? MAX_SAFE_FLOAT : props.maxValue;
+    minValue = isNaN(props.minValue) ? MIN_SAFE_FLOAT : props.maxValue;
+  }
+
+
+  let {minusSign, plusSign} = numberParser.symbols;
+
+  // javascript doesn't recognize NaN === NaN, so multiple onChanges will get fired if we don't ignore consecutive ones
+  // in addition, if the input starts with a number, then we'll count that as the last val dispatched, we only need to calculate it the first time
+  let startingValue = useMemo(() => {
+    if (!isNaN(value)) {
+      return value;
+    } else if (!isNaN(defaultValue)) {
+      return defaultValue;
+    }
+    return NaN;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let lastValDispatched = useRef(startingValue);
+  let smartOnChange = useCallback((val) => {
+    if (!isNaN(val) || !isNaN(lastValDispatched.current)) {
+      onChange?.(val);
+    }
+    lastValDispatched.current = val;
+  }, [lastValDispatched, onChange]);
 
-  const numberParser = useNumberParser();
-  const textValueFormatter = useNumberFormatter(formatOptions);
-  const inputValueFormatter = useNumberFormatter();
 
-  const [numberValue, setNumberValue] = useControlledState<number>(value, defaultValue || 0, onChange);
-  let initialInputValue = inputValueFormatter.format(numberValue);
-  const [inputValue, setInputValue] = useState(isNaN(value) && isNaN(defaultValue) ? '' : initialInputValue);
-  const [isValid, setIsValid] = useState(isInputValueValid(numberValue, maxValue, minValue));
+  let [numberValue, setNumberValue] = useControlledState<number>(value, isNaN(defaultValue) ? NaN : defaultValue, smartOnChange);
 
-  const minusSign = useRef('-');
+  let textValue = inputValueFormatter.format(numberValue);
+  let [inputValue, setInputValue] = useState(() => isNaN(numberValue) ? '' : textValue);
 
+
+  // this updates the field only if the formatter or number has changed,
+  // this should only run after increment, decrement, blur
+  // or when the formatter changes, but not because of the overrides, which is why we run on formatOptions
+  // I doubt that the formatOptions will change while a user is typing a number, but that might cause a problem
   useEffect(() => {
-    // Get the minus sign of the current locale to filter the input value
-    // Automatically updates the minus sign when numberFormatter changes
-    minusSign.current = inputValueFormatter.formatToParts(-11).find(p => p.type === 'minusSign').value;
-  }, [inputValueFormatter]);
+    setInputValue(isNaN(numberValue) ? '' : textValue);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [formatOptions, setInputValue, numberValue, locale]);
+
+  let parse = (value: string): number => {
```

Comment:
Feels like most of this wrapper should also be part of the actual number parser itself?

Reply:
I don't quite follow what you're asking for